### PR TITLE
feat(extension): redirect by geolocation 

### LINKED ISSUE #11: feat: Access Origin by geolocation with 302 redirection
After discussing with  @minche-tsai , we are interested to implement the **Access Origin by geolocation** extension.

圖片

### Primary Use Case

1. Dynamically route viewer requests to different S3 buckets in different regions(see this [blog post](https://aws.amazon.com/tw/blogs/networking-and-content-delivery/dynamically-route-viewer-requests-to-any-origin-using-lambdaedge/))

### Other Use Cases
1. Two API Gateways with one in `ap-northeast-1` and the other in `us-west-2` running the same serverless application sharing the same cloudfront distribution. Users visiting the PoP should be routed to nearest origin.
2. One in `cn-north-1` and the other in `ap-northeast-1`, all china clients visiting the global CF edges should go to `cn-north-1` while others `ap-northeas-1`.

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by minche-tsai**
Fixes #11

### CODE REVIEW

File: API.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 Name|Description
 ----|-----------
+[AccessOriginByGeolocation](#cdk-cloudfront-plus-accessoriginbygeolocation)|*No description*
```
Comment (guan840912) [inline API.md:7]: add some description to this class ??
cc @pahud @minche-tsai

File: src/extensions.ts [hunk 2]
```diff
@@ -313,6 +313,35 @@ export class SelectOriginByViwerCountry extends Custom {
...
 };
 
+export interface AccessOriginByGeolocationProps {
+  /**
+   * The pre-defined country code table.
+   * Exampe: { 'US': 'amazon.com' }
+   */
+  readonly countryTable: { [code: string]: string };
+}
+
```
Comment (guan840912) [inline src/extensions.ts:323]: add some description cc @pahud  @MinCheTsai  
like https://github.com/pahud/cdk-cloudfront-plus/blob/d611a6b6167be7682418ac9277b61331ebdf4c55/src/extensions.ts#L263-L267

**Review by guan840912 — APPROVED** [review]

File: src/extensions.ts (hunk 2 above)
Comment (minche-tsai) [reply src/extensions.ts:323]: updated

File: API.md [hunk 3]
```diff
@@ -35,6 +37,30 @@ Name|Description
...
+__Implements__: [IConstruct](#constructs-iconstruct), [IConstruct](#aws-cdk-core-iconstruct), [IConstruct](#constructs-iconstruct), [IDependable](#aws-cdk-core-idependable), [ITaggable](#aws-cdk-core-itaggable), [IExtensions](#cdk-cloudfront-plus-iextensions)
+__Extends__: [Custom](#cdk-cloudfront-plus-custom)
+
+### Initializer
+
+
+
+
+```ts
+new AccessOriginByGeolocation(scope: Construct, id: string, props: AccessOriginByGeolocationProps)
```
Comment (pahud) [inline API.md:53]:
```suggestion
new RedirectByGeolocation(scope: Construct, id: string, props: RedirectByGeolocationProps)
```
Let's rename to `RedirectByGeolocation`

File: lambda-assets/extensions/cf-access-origin-by-geolocation/index.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
+export async function handler(event: any) {
```
Comment (pahud) [inline lambda-assets/extensions/cf-access-origin-by-geolocation/index.ts:1]: Let's rename this file to lambda-assets/extensions/cf-redirect-by-geolocation/index.ts

Thanks

File: src/__tests__/integ.access-origin-by-geolocation.test.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import { SynthUtils } from '@aws-cdk/assert';
```
Comment (pahud) [inline src/__tests__/integ.access-origin-by-geolocation.test.ts:1]: let's rename to `integ.redirect-by-geolocation.test.ts`

File: src/demo/access-origin-by-geolocation/index.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+import * as cf from '@aws-cdk/aws-cloudfront';
+import * as origins from '@aws-cdk/aws-cloudfront-origins';
+import * as cdk from '@aws-cdk/core';
+import * as extensions from '../../extensions';
+
+const app = new cdk.App();
+
+const stack = new cdk.Stack(app, 'access-origin-by-geolocation');
```
Comment (pahud) [inline src/demo/access-origin-by-geolocation/index.ts:8]:
```suggestion
const stack = new cdk.Stack(app, 'redirect-by-geolocation');
```

File: src/demo/access-origin-by-geolocation/index.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+// create the cloudfront distribution with extension(s)
+const ext = new extensions.AccessOriginByGeolocation(stack, 'AccessOriginByGeolocation', {
```
Comment (pahud) [inline src/demo/access-origin-by-geolocation/index.ts:11]:
```suggestion
const ext = new extensions.RedirectByGeolocation(stack, 'RedirectByGeolocation', {
```

File: src/extensions.ts [hunk 8]
```diff
@@ -313,6 +313,38 @@ export class SelectOriginByViwerCountry extends Custom {
...
+      bundling: {
+        define: {
+          'process.env.COUNTRY_CODE_TABLE': jsonStringifiedBundlingDefinition(props.countryTable),
+        },
+      },
+    });
+    super(scope, id, {
+      func: func,
+      eventType: cf.LambdaEdgeEventType.ORIGIN_REQUEST,
+      solutionId: '',
```
Comment (pahud) [inline src/extensions.ts:342]:
```suggestion
      solutionId: '8135',
```

Please use this solutionId thanks.

File: src/extensions.ts [hunk 9]
```diff
@@ -313,6 +313,38 @@ export class SelectOriginByViwerCountry extends Custom {
...
+      templateDescription: 'Cloudfront extension with AWS CDK - access Origin by geolocation',
```
Comment (pahud) [inline src/extensions.ts:343]:
```suggestion
      templateDescription: 'Cloudfront extension with AWS CDK - Redirect by Geolocation',
```

**Review by pahud — CHANGES_REQUESTED** [review]
Hi @minche-tsai  some minor changes please

1. The extension name is confirmed to be `RedirectByGeolocation` please rename all filenames and class name accordingly.
2.  The solutionId is provided in the review, please just commit the suggested changes.

Everything else LGTM. Thanks!

File: src/extensions.ts (hunk 8 above)
Comment (guan840912) [reply src/extensions.ts:342]: SO8135 or 8135 ??

`DefaultDirIndex` use `SO8134`
https://github.com/pahud/cdk-cloudfront-plus/blob/d611a6b6167be7682418ac9277b61331ebdf4c55/src/extensions.ts#L256

**Comment by minche-tsai** [comment]
> Hi @minche-tsai some minor changes please
> 
> 1. The extension name is confirmed to be `RedirectByGeolocation` please rename all filenames and class name accordingly.
> 2. The solutionId is provided in the review, please just commit the suggested changes.
> 
> Everything else LGTM. Thanks!

Rename done

**Review by pahud — APPROVED** [review]
:shipit:

**Pull request merged**